besspin/cyberPhys/cyberphyslib/cyberphyslib/demonstrator/health.py

```diff
@@ -0,0 +1,334 @@
+        self.heartbeat_thread.on_poll = self.on_heartbeat_poll
+        self.heartbeat_thread.on_start = self.on_heartbeat_start
+        self.heartbeat_thread.on_exit = self.on_heartbeat_exit
+        self._heartbeat_call_started = False
+
+    def on_start(self):
+        super().on_start()
+        self.heartbeat_thread.start()
+
+    def on_exit(self):
+        self.heartbeat_thread.exit()
+        super().exit()
+
+    def on_heartbeat_start(self):
+        pass
+
+    def on_heartbeat_exit(self):
+        pass
+
+    def on_heartbeat_poll(self, t):
+        pass
+
+
+class HeartbeatClientComponent(HeartbeatTaskComponent):
+    """
+    Heartbeat Client (component to be monitored)
+
+    Examples:
+        ```
+        myc = heartbeat_component("myc", {("msg-in", 5006)}, {"msg-out", 5007})
+        my_id = 0x45
+        myc.register_heartbeat_bus(bus_obj, my_id)
+        myc.start()
+        ```
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(HeartbeatClientComponent, self).__init__(*args, **kwargs)
+        self._heartbeat_client = None
+
+    def register_heartbeat_bus(self, bus: TcpBus, client_id: int):
+        """register heartbeat client component with TCP bus and a client identifier"""
+        self._heartbeat_client: TcpHeartbeatClient = TcpHeartbeatClient(client_id, bus)
+
+    def on_heartbeat_poll(self, t):
+        if self._heartbeat_client is not None:
+            r = self._heartbeat_client.recv(1.0)
+            while not isinstance(r, can.Message) or r.arbitration_id != canspecs.CAN_ID_HEARTBEAT_REQ:
+                r = self._heartbeat_client.recv()
+            idx = struct.unpack("!I", r.data)[0]
+            self._heartbeat_client.send_heartbeat_ack_msg(idx)
+
+
+class HeartbeatMonitorComponent(HeartbeatTaskComponent):
+    """
+    Heartbeat Monitor Component (component to check the health of the clients)
+
+    FIXME: all clients must be registered or else the monitor wont receive at the correct rate
+
+    Examples:
+        ```
+        myc = heartbeat_component("myc", {("msg-in", 5006)}, {"msg-out", 5007})
+        client_ids = [0x45, 0x56]
+        myc.register_heartbeat_bus(bus_obj, client_ids)
+        myc.start()
+        ```
+    """
+    def __init__(self, *args, **kwargs):
+        super(HeartbeatMonitorComponent, self).__init__(*args, **kwargs)
+        self._heartbeat_monitor = None
+        self.n_clients = 0
+
+    def register_heartbeat_bus(self, bus: TcpBus, client_ids):
+        """register the monitor on a bus and provide the clients to be monitored"""
+        self._heartbeat_monitor: TcpHeartbeatMonitor = TcpHeartbeatMonitor(client_ids, bus)
+        self.n_clients = len(client_ids)
+
+    def on_heartbeat_poll(self, t):
+        if self._heartbeat_monitor is not None:
+            hm: TcpHeartbeatMonitor = self._heartbeat_monitor
+            hm.send_req_nodes()
+            rets = [self._heartbeat_monitor.tcp_bus.recv(1.0) for _ in range(self.n_clients)]
```

Comment:
That looks reasonable. The recv timeout should be configurable (a class variable).